File: algorithm/C++/1143-Longest_Common_Subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int m = text1.size();
        int n = text2.size();
        vector<vector<int>> table(m + 1, vector<int>(n + 1, 0));
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (text1[i] == text2[j]) {
                    table[i + 1][j + 1] = table[i][j] + 1;
                } else {
                    table[i + 1][j + 1] = max(table[i + 1][j], table[i][j + 1]);
                }
            }
        }
        return table[m][n];
    }
};
// Time : O(mn)
// Space : O(mn)

File: algorithm/Go/1143-Longest_Common_Subsequence.go
func longestCommonSubsequence(text1 string, text2 string) int {
    s1 := len(text1)
    s2 := len(text2)
    table := make([][]int, s1 + 1)
    for i:=0; i<=s1; i++ {
        table[i] = make([]int, s2 + 1)
    }
    for i:=0; i<=len(text1); i++ {
        for j:=0; j<=len(text2); j++ {
            table[i][j] = 0
        }
    }
    maxLen := 0
    for i:=0; i<len(text1); i++ {
        for j:=0; j<len(text2); j++ {
            if text1[i] == text2[j] {
                table[i + 1][j + 1] = max(table[i + 1][j + 1], table[i][j] + 1)
            } else {
                table[i + 1][j + 1] = max(table[i + 1][j], table[i][j + 1])
            }
            maxLen = max(maxLen, table[i + 1][j + 1])
        }
    }
    return maxLen
}
func max(i int, j int) int {
    if i < j {
        return j
    }
    return i
}
// Time : O(n^2)
// Space : O(n^2)